src/hooks/useManifoldBoardBuilder.ts

```diff
@@ -90,13 +93,129 @@ export const useManifoldBoardBuilder = (
+  }, [circuitJson])
+
   const boardData = useMemo(() => {
     const boards = su(circuitJson).pcb_board.list()
-    if (boards.length === 0) {
-      // Error will be set in effect
-      return null
+    if (boards.length > 0) {
+      return boards[0]!
     }
-    return boards[0]!
+
+    // Check if there are components or holes that need a faux board
+    const cadComponents = su(circuitJson).cad_component.list()
+    const holes = su(circuitJson).pcb_hole.list()
+    const platedHoles = su(circuitJson).pcb_plated_hole.list()
+    const vias = su(circuitJson).pcb_via.list()
+
+    if (
+      cadComponents.length > 0 ||
+      holes.length > 0 ||
+      platedHoles.length > 0 ||
+      vias.length > 0
+    ) {
+      // Calculate bounding box of components and holes
+      let minX = Infinity
+      let maxX = -Infinity
+      let minY = Infinity
+      let maxY = -Infinity
+
+      cadComponents.forEach((component: CadComponent) => {
+        if (component.position) {
+          const x = component.position.x
+          const y = component.position.y
+          minX = Math.min(minX, x)
+          maxX = Math.max(maxX, x)
+          minY = Math.min(minY, y)
+          maxY = Math.max(maxY, y)
+        }
+      })
+
+      // Include holes in bounding box
+      holes.forEach((hole) => {
+        const x = hole.x
+        const y = hole.y
+        const radius =
+          hole.hole_shape === "circle"
+            ? hole.hole_diameter / 2
+            : hole.hole_shape === "pill"
+              ? Math.max(hole.hole_width, hole.hole_height) / 2
+              : hole.hole_shape === "rotated_pill"
+                ? Math.max(hole.hole_width, hole.hole_height) / 2
+                : 0
+        minX = Math.min(minX, x - radius)
+        maxX = Math.max(maxX, x + radius)
+        minY = Math.min(minY, y - radius)
+        maxY = Math.max(maxY, y + radius)
+      })
+
+      // Include plated holes
+      platedHoles.forEach((hole) => {
+        const x = hole.x
+        const y = hole.y
+        let radius = 0
+        if (hole.shape === "circle") {
+          radius = hole.hole_diameter / 2
+        } else if (hole.shape === "oval" || hole.shape === "pill") {
+          radius = Math.max(hole.hole_width, hole.hole_height) / 2
+        }
+        minX = Math.min(minX, x - radius)
+        maxX = Math.max(maxX, x + radius)
+        minY = Math.min(minY, y - radius)
+        maxY = Math.max(maxY, y + radius)
+      })
+
+      // Include vias
+      vias.forEach((via) => {
+        const x = via.x
+        const y = via.y
+        const radius = via.outer_diameter / 2
+        minX = Math.min(minX, x - radius)
+        maxX = Math.max(maxX, x + radius)
+        minY = Math.min(minY, y - radius)
+        maxY = Math.max(maxY, y + radius)
+      })
+
+      // If no valid positions found, use default
+      if (minX === Infinity) {
+        minX = -10
+        maxX = 10
+        minY = -10
+        maxY = 10
+      }
```

Comment:
this kind of logic is so so messy to put inline in a hook. If you're adding 100+ lines to a closure you are making sloppy code and you should have a new function and import instead